settings/settings.txt
{"classType_of_environment": {"py/type": "Environments.Environment2D.Environment2D"}, "classType_of_population": {"py/type": "CPopulation.CPopulation"}, "classType_settings": {"classType_of_environment": {"height": 800, "width": 800, "itemSpeed": 7, "itemSize": 7}, "classType_of_population": {"maximal_number_of_saved_couples": 1000, "set_initial_position_in_the_environment": null, "a": 0.5, "type_of_average": 0, "classType_of_male_individual": {"py/type": "CIndividual.CIndividual"}, "classType_of_female_individual": {"py/type": "CIndividual.CIndividual"}, "sex_ratio": 0.5, "population_size": 500, "classType_settings": {"classType_of_male_individual": {"father": null, "mother": null, "latency": 0.2, "classType_settings": {"classType_of_chromosome": {"loci": [{"py/tuple": [0, false]}, {"py/tuple": [0.5, true]}, {"py/tuple": [0, false]}, {"py/tuple": [0.5, true]}], "mutation_range": 0.01, "mutation_rate": 0.01}}, "gender": 1, "survival_prob": 0.999, "classType_of_chromosome": {"py/type": "CChromosome.CChromosome"}}, "classType_of_female_individual": {"father": null, "mother": null, "latency": 0.8, "classType_settings": {"classType_of_chromosome": {"loci": [{"py/tuple": [0, false]}, {"py/tuple": [0.5, true]}, {"py/tuple": [0, false]}, {"py/tuple": [0.5, true]}], "mutation_range": 0.01, "mutation_rate": 0.01}}, "gender": 0, "survival_prob": 0.999, "classType_of_chromosome": {"py/type": "CChromosome.CChromosome"}}}}}}

settings/default_settings.txt
{"classType_of_population": [{"py/tuple": [{"py/type": "CPopulation.CPopulation"}, {"classType_of_female_individual": [{"py/tuple": [{"py/type": "CIndividual.CIndividual"}, {"father": {"v": "null", "py/object": "__main__.Reference"}, "latency": {"v": "0.8", "py/object": "__main__.Reference"}, "gender": {"v": "0", "py/object": "__main__.Reference"}, "survival_prob": {"v": "0.999", "py/object": "__main__.Reference"}, "classType_of_chromosome": [{"py/tuple": [{"py/type": "CChromosome.CChromosome"}, {"mutation_rate": {"v": "0.01", "py/object": "__main__.Reference"}, "mutation_range": {"v": "0.01", "py/object": "__main__.Reference"}, "loci": {"v": "[{\"py/tuple\": [0, false]}, {\"py/tuple\": [0.5, true]}, {\"py/tuple\": [0, false]}, {\"py/tuple\": [0.5, true]}]", "py/object": "__main__.Reference"}}]}], "mother": {"v": "null", "py/object": "__main__.Reference"}}]}], "sex_ratio": {"v": "0.5", "py/object": "__main__.Reference"}, "population_size": {"v": "5000", "py/object": "__main__.Reference"}, "set_initial_position_in_the_environment": {"v": "null", "py/object": "__main__.Reference"}, "classType_of_male_individual": [{"py/tuple": [{"py/type": "CIndividual.CIndividual"}, {"father": {"v": "null", "py/object": "__main__.Reference"}, "latency": {"v": "0.2", "py/object": "__main__.Reference"}, "gender": {"v": "1", "py/object": "__main__.Reference"}, "survival_prob": {"v": "0.999", "py/object": "__main__.Reference"}, "classType_of_chromosome": [{"py/tuple": [{"py/type": "CChromosome.CChromosome"}, {"mutation_rate": {"v": "0.01", "py/object": "__main__.Reference"}, "mutation_range": {"v": "0.01", "py/object": "__main__.Reference"}, "loci": {"v": "[{\"py/tuple\": [0, false]}, {\"py/tuple\": [0.5, true]}, {\"py/tuple\": [0, false]}, {\"py/tuple\": [0.5, true]}]", "py/object": "__main__.Reference"}}]}], "mother": {"v": "null", "py/object": "__main__.Reference"}}]}], "type_of_average": {"v": "0", "py/object": "__main__.Reference"}, "maximal_number_of_saved_couples": {"v": "1000", "py/object": "__main__.Reference"}, "a": {"v": "0.5", "py/object": "__main__.Reference"}}]}, {"py/tuple": [{"py/type": "CPopulation.CPopulationImmediateFlush"}, {"classType_of_female_individual": [{"py/tuple": [{"py/type": "CIndividual.CIndividual"}, {"father": {"v": "null", "py/object": "__main__.Reference"}, "latency": {"v": "0.8", "py/object": "__main__.Reference"}, "gender": {"v": "0", "py/object": "__main__.Reference"}, "survival_prob": {"v": "0.999", "py/object": "__main__.Reference"}, "classType_of_chromosome": [{"py/tuple": [{"py/type": "CChromosome.CChromosome"}, {"mutation_rate": {"v": "0.01", "py/object": "__main__.Reference"}, "mutation_range": {"v": "0.01", "py/object": "__main__.Reference"}, "loci": {"v": "[{\"py/tuple\": [0, false]}, {\"py/tuple\": [0.5, true]}, {\"py/tuple\": [0, false]}, {\"py/tuple\": [0.5, true]}]", "py/object": "__main__.Reference"}}]}], "mother": {"v": "null", "py/object": "__main__.Reference"}}]}], "sex_ratio": {"v": "0.5", "py/object": "__main__.Reference"}, "population_size": {"v": "5000", "py/object": "__main__.Reference"}, "set_initial_position_in_the_environment": {"v": "null", "py/object": "__main__.Reference"}, "classType_of_male_individual": [{"py/tuple": [{"py/type": "CIndividual.CIndividual"}, {"father": {"v": "null", "py/object": "__main__.Reference"}, "latency": {"v": "0.2", "py/object": "__main__.Reference"}, "gender": {"v": "1", "py/object": "__main__.Reference"}, "survival_prob": {"v": "0.999", "py/object": "__main__.Reference"}, "classType_of_chromosome": [{"py/tuple": [{"py/type": "CChromosome.CChromosome"}, {"mutation_rate": {"v": "0.01", "py/object": "__main__.Reference"}, "mutation_range": {"v": "0.01", "py/object": "__main__.Reference"}, "loci": {"v": "[{\"py/tuple\": [0, false]}, {\"py/tuple\": [0.5, true]}, {\"py/tuple\": [0, false]}, {\"py/tuple\": [0.5, true]}]", "py/object": "__main__.Reference"}}]}], "mother": {"v": "null", "py/object": "__main__.Reference"}}]}], "type_of_average": {"v": "0", "py/object": "__main__.Reference"}, "maximal_number_of_saved_couples": {"v": "5000", "py/object": "__main__.Reference"}, "a": {"v": "0.5", "py/object": "__main__.Reference"}}]}], "classType_of_environment": [{"py/tuple": [{"py/type": "Environments.Environment2D.Environment2D"}, {"itemSpeed": {"v": "7", "py/object": "__main__.Reference"}, "itemSize": {"v": "7", "py/object": "__main__.Reference"}, "width": {"v": "800", "py/object": "__main__.Reference"}, "height": {"v": "800", "py/object": "__main__.Reference"}}]}, {"py/tuple": [{"py/type": "Environments.Environment2D.Environment2DNoBounce"}, {"itemSpeed": {"v": "7", "py/object": "__main__.Reference"}, "itemSize": {"v": "7", "py/object": "__main__.Reference"}, "width": {"v": "800", "py/object": "__main__.Reference"}, "height": {"v": "800", "py/object": "__main__.Reference"}}]}, {"py/tuple": [{"py/type": "Environments.EnvironmentSoup.EnvironmentSoup"}, {"e": {"v": "0.999", "py/object": "__main__.Reference"}}]}, {"py/tuple": [{"py/type": "Environments.Environment2DFaster.Environment2DDelaunay"}, {"itemSpeed": {"v": "7", "py/object": "__main__.Reference"}, "itemSize": {"v": "7", "py/object": "__main__.Reference"}, "width": {"v": "800", "py/object": "__main__.Reference"}, "height": {"v": "800", "py/object": "__main__.Reference"}}]}]}

settings/settings.py
from Tools.usefulDecorators import printAllParameters
import jsonpickle

__author__ = 'robert'

"""
The settings of the simulation are saved as a dictionary. The keys of the dictionary should correspond to the parameter-
names of the chosen classes of the environment, individual, population and chromosomes.

Examples:
1) The following settings choose for the population the class CPopulation.CPopulationImmediateFlush, for the environment
the class EnvironmentSoup.EnvironmentSoup

        self.settings_dict = {}
        self.settings_dict['classType_of_population'] = CPopulation.CPopulationImmediateFlush
        self.settings_dict['population_settings'] = {}
        self.settings_dict['population_settings']['population_size'] = 5000
        self.settings_dict['population_settings']['maximal_number_of_saved_couples'] = \
            2*self.settings_dict['population_settings']['population_size']
        self.settings_dict['population_settings']['sex_ratio'] = 0.5
        self.settings_dict['population_settings']['type_of_average'] = CPopulation.ARITHMETIC_MEAN
        self.settings_dict['population_settings']['a'] = 0.5
        self.settings_dict['population_settings']['male_individual_settings'] = {}
        self.settings_dict['population_settings']['male_individual_settings']['gender'] = CIndividual.MALE
        self.settings_dict['population_settings']['male_individual_settings']['survival_prob'] = 0.99
        self.settings_dict['population_settings']['male_individual_settings']['latency'] = 0.98
        self.settings_dict['population_settings']['male_individual_settings']['chromosome_settings'] = {}
        self.settings_dict['population_settings']['male_individual_settings']['chromosome_settings']['mutation_range'] = 0.01
        self.settings_dict['population_settings']['male_individual_settings']['chromosome_settings']['mutation_rate'] = 0.01
        self.settings_dict['population_settings']['male_individual_settings']['chromosome_settings']['loci']\
            = [(0, False), (0.5, True), (0, False), (0.5, True)] # initial_chromosome

        self.settings_dict['population_settings']['female_individual_settings'] = {}
        self.settings_dict['population_settings']['female_individual_settings']['gender'] = CIndividual.FEMALE
        self.settings_dict['population_settings']['female_individual_settings']['survival_prob'] = 0.99
        self.settings_dict['population_settings']['female_individual_settings']['latency'] = 0.99
        self.settings_dict['population_settings']['female_individual_settings']['chromosome_settings'] = {}
        self.settings_dict['population_settings']['female_individual_settings']['chromosome_settings']['mutation_range'] = 0.01
        self.settings_dict['population_settings']['female_individual_settings']['chromosome_settings']['mutation_rate'] = 0.01
        self.settings_dict['population_settings']['female_individual_settings']['chromosome_settings']['loci']\
            = [(0, False), (0.5, True), (0, False), (0.5, True)] # initial_chromosome

        self.settings_dict['classType_of_environment'] = EnvironmentSoup.EnvironmentSoup
        self.settings_dict['environment_settings'] = {}
        self.settings_dict['environment_settings']['e'] = 0.99

2) The following settings choose for the environment the class Environment2DDelaunay.Environment2DDelaunay

        # insert population settings from up

        # You can choose with the same settings also Environment2D.Environment2D or Environment2D.Environment2DNoBounce
        self.settings_dict['classType_of_environment'] = Environment2D.Environment2D
        self.settings_dict['environment_settings'] = {}
        self.settings_dict['environment_settings']['width'] = 800
        self.settings_dict['environment_settings']['height'] = 800
        self.settings_dict['environment_settings']['itemSize'] = 6 # size_of_individuals
        self.settings_dict['environment_settings']['itemSpeed'] = 7 # speed_of_individuals
"""

class CSimulationData:
    """
    Stores the settings of the simulation
    """
    def __init__(self, settings_file):
        self.collision_counter = 0
        self.step_counter = 1
        self.average_number_of_collisions_per_timestep = 0
        if settings_file is not None:
            self.load_settings_from_file(settings_file)

    @printAllParameters
    def __str__(self):
        return ""

    def save_settings_to_file(self, filename):
        """
        Converts the settings into json-format and saves it in the given file
        :param filename: 'relative path/filename'
        """
        settings_as_json = jsonpickle.encode(self.settings_dict)
        f = open(filename, 'w')
        f.write(settings_as_json)
        print("settings saved to "+filename)

    def load_settings_from_file(self, filename):
        f = open(filename)
        json_str = f.read()
        self.settings_dict = jsonpickle.decode(json_str)
        print("settings loaded from "+filename)